### ISSUE TITLE
MonoGame is not AppDomain-friendly

### ISSUE THREAD

**Issue by kameko** [mono/MonoGame#4787]
Attempting to pass MonoGame types through a MarshalByRefObject into another AppDomain crashes, since no MonoGame types are marked as Serializable. I don't know if I'm just being an idiot and don't know how AppDomains work, but I can't find a way around this. I don't even know if this is a concern for the project, but just letting you know.

**Comment by totallyeviljake** [comment]
Serialization is a copy process for an instance. Marshalling by reference is when a proxy object maintains a connection to the original memory space. MonoGame is not really good for that, but it can be done. You only serialize when you want to copy the memory space and when you can deserialize back to the original object's memory space.

**Comment by totallyeviljake** [comment]
So, if you try to marshal an object like a framebuffer then you may well get into memory corruption. So you would want to marshal your own proxy object that interacts with the framebuffer. 

Some instance data types should be serializable in MG so that you can pass their value across the marshal boundary. That's how much of the "load onto the framebuffer" code works. You just don't see it.

**Comment by kameko** [comment]
I'm mostly talking about the simpler datatypes such as Vector2 and Texture2D not being marshalable. It's difficult when not even those can be passed around.

**Comment by totallyeviljake** [comment]
Vector2 = serializable data object. never intended for marshalling
Texture2D = very complex type that should never be marshal'd. It contains a handle to framebuffer resources that can not be shared across memory boundaries.

**Comment by kameko** [comment]
Darn, I figured Texture2D was like that. That's why I gave up marking stuff as serializable in MonoGame's source when the stack trace told me that, haha.
Well, I guess this is an unsolvable problem then? I suppose if the developers think so then they can close this.

**Comment by KonajuGames** [comment]
What is the end goal of what you are trying to achieve?

**Comment by kameko** [comment]
I was isolating modules in AppDomains, so, say, a game server could send a DLL to an end-user that would be loaded in an AppDomain with no permissions. The DLL in the AppDomain could be unloaded while the engine is running, and it wouldn't be able to do anything malicious.
I've since abandoned AppDomains and I'm just using a scripting language instead, but it would be a nice feature for people who like to be vanilla with as few dependencies as possible. Before I abandoned that prospect, I was in the midst of writing a convoluted "remote management system" that could pass simple native datatypes to control a database of MonoGame types. It was such a mess compared to using a scripting language that I just gave up.
